encryption: Limit encryption keys to 2 bits

tgraf opened the pull request.

joestringer commented on bpf/lib/common.h:
```diff
@@ -453,6 +453,29 @@ enum {
+ * M M M M
+ * (see MARK_MAGIC_* in bpf/lib/common.h)
+ *
+ * Encryption ID (2 bits):
+ * E E  Key ID
```
As I noted out-of-band while Github was down, the 0x2000 bit will collide with portmap CNI so we still have that overlap. For that I think we're fine to just state an incompatibility between combining ipsec with portmap CNI (suggesting users to instead use kube-proxy-replacement). [Reference](https://github.com/fwmark/registry/#bitwise-mark-registry)

tgraf replied:
Added

joestringer commented on bpf/lib/common.h:
```diff
@@ -453,6 +453,29 @@ enum {
+ *
+ * Cilium Mark (4 bits):
+ * M M M M
+ * (see MARK_MAGIC_* in bpf/lib/common.h)
```
nit:
```suggestion
 * (see MARK_MAGIC_* below)
```

joestringer commented on .github/workflows/encryption.yaml:
```diff
@@ -0,0 +1,103 @@
+
+      - name: Install Cilium CLI
+        run: |
+          curl -LO https://github.com/cilium/cilium-cli/releases/download/v0.4/cilium-linux-amd64.tar.gz
```
Is this deliberately set to `v0.4` or should we consider `latest` here?

Put another way, do we want to update this regularly and how? Bump every time there's a Cilium release? Bump every time there's a cilium-cli release?

tgraf replied:
I've removed the CI test for now from this PR as it will be more involved to get it into a useful state.

joestringer commented on .github/workflows/encryption.yaml:
```diff
@@ -0,0 +1,103 @@
+          cilium connectivity test
+
+      - name: Cleanup
+        if: ${{ always() }}
```
Do we always collect a sysdump? Or should this be the following?
```suggestion
        if: ${{ failure() }}
```

tgraf replied:
I've removed the CI test for now from this PR as it will be more involved to get it into a useful state.

joestringer commented on .github/workflows/encryption.yaml:
```diff
@@ -0,0 +1,103 @@
+        run: |
+          cilium connectivity test
+
+      - name: Cleanup
```
nit: This doesn't seem to be cleaning up, seems more like:

```suggestion
      - name:  Capture cilium-sysdump
```

tgraf replied:
I've removed the CI test for now from this PR as it will be more involved to get it into a useful state.

joestringer reviewed (changes requested):
All minor nits, would be also good to get @jrfastab 's eyes on the rotation bit.

joestringer commented on bpf/lib/common.h:
```diff
@@ -463,7 +486,13 @@ enum {
+ *  - the key index to use for encryption when multiple keys are in-flight.
+ *    In the IPsec case this becomes the SPI on the wire.
+ *
+ * The 1 bit at 0x2000 is used by portmap
```
nit:
```suggestion
 * The 1 bit at 0x2000 is used by portmap CNI in chaining mode. It is not used by the eBPF implementation of portmap.
```

joestringer reviewed (approved):
One minor clarification.

nbusseneau reviewed (approved):
LGTM.

qmonnet commented on Documentation/operations/upgrade.rst:
```diff
@@ -335,6 +335,16 @@ Annotations:
+   Please review the following changes before upgrading to 1.10 as some of the
+   changes may require an action in your cluster before upgrading.
+
+* The encryption key ID has been limited to 3 bits (stored the skb mark) in
```
```suggestion
* The encryption key ID has been limited to 2 bits (stored in the skb mark) in
```

brb commented on bpf/lib/common.h:
```diff
@@ -453,6 +453,29 @@ enum {
 #define METRIC_SERVICE  3
 
 /* Magic ctx->mark identifies packets origination and encryption status.
+ *
```
:heart: this comment.

tgraf committed:
encryption: Limit encryption keys to 2 bits

The upper 2 bits of the encryption key stored in the skb->mark currently
clash with kube-proxy usage. It was believed that packets with these
marks would never be exposed to the respective iptables chains but with
recent changes in the datapath, this is no longer true.

There is not really a reason to use more than 2 bits. As it is uncommon
to have more than 2 keys in play. Use of 2 bits allows to use 3 keys
plus 0 as the no key indication.

Adjust datapath, validation, and docs.

Signed-off-by: Thomas Graf <thomas@cilium.io>
